src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -25,22 +25,712 @@
+            : m_includeFinder(_preprocessOptions.includeFinder), m_logger(_preprocessOptions.logger), m_pragmaStage(IShader::ESS_UNKNOWN) {}
+
+        const IShaderCompiler::CIncludeFinder* m_includeFinder;
+        system::logger_opt_ptr m_logger;
+        IShader::E_SHADER_STAGE m_pragmaStage;
+
+
+        template <typename ContextT>
+        bool locate_include_file(ContextT& ctx, std::string& file_path, bool is_system, char const* current_name, std::string& dir_path, std::string& native_name) 
+        {
+            //on builtin return true
+            dir_path = ctx.get_current_directory().string();
+            std::optional<std::string> result;
+            if (is_system) {
+                result = m_includeFinder->getIncludeStandard(dir_path, file_path);
+                dir_path = "";
+            }
+            else
+                result = m_includeFinder->getIncludeRelative(dir_path, file_path);
+            if (!result)
+            {
+                m_logger.log("Pre-processor error: Bad include file.\n'%s' does not exist.", nbl::system::ILogger::ELL_ERROR, file_path.c_str());
+                return false;
+            }
+            native_name = file_path;
+            return true;
+        }
+
+
+        // interpretation of #pragma's of the form 'wave option[(value)]'
+        template <typename ContextT, typename ContainerT>
+        bool
+            interpret_pragma(ContextT const& ctx, ContainerT& pending,
+                typename ContextT::token_type const& option, ContainerT const& values,
+                typename ContextT::token_type const& act_token)
+        {
+            auto optionStr = option.get_value().c_str();
+            if (strcmp(optionStr, "shader_stage") == 0) 
+            {
+                auto valueIter = values.begin();
+                if (valueIter == values.end()) {
+                    m_logger.log("Pre-processor error:\nMalformed shader_stage pragma. No shaderstage option given", nbl::system::ILogger::ELL_ERROR);
+                    return false;
+                }
+                auto shaderStageIdentifier = std::string(valueIter->get_value().c_str());
+                core::unordered_map<std::string, IShader::E_SHADER_STAGE> stageFromIdent = {
+                    { "vertex", IShader::ESS_VERTEX },
+                    { "fragment", IShader::ESS_FRAGMENT },
+                    { "tesscontrol", IShader::ESS_TESSELLATION_CONTROL },
+                    { "tesseval", IShader::ESS_TESSELLATION_EVALUATION },
+                    { "geometry", IShader::ESS_GEOMETRY },
+                    { "compute", IShader::ESS_COMPUTE }
+                };
+                auto found = stageFromIdent.find(shaderStageIdentifier);
+                if (found == stageFromIdent.end())
+                {
+                    m_logger.log("Pre-processor error:\nMalformed shader_stage pragma. Unknown stage '%s'", nbl::system::ILogger::ELL_ERROR, shaderStageIdentifier);
```

Comment:
Remember to use `c_str()` for the `%s` format argument!

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/608#discussion_r1399322577